lib/install/rust.js

```diff
@@ -0,0 +1,557 @@
+var request = require('request');
+var tags = require('common-tags');
+var tar = require('tar-fs');
+
+// Internal
+var log = require('../log');
+
+var SDK_PATHS = {
+  sdk: path.join(osenv.home(), '.tessel/sdk'),
+  rustlib: path.join(osenv.home(), '.tessel/rust'),
+};
+
+var SDK_URLS = {
+  macos: 'https://builds.tessel.io/t2/sdk/t2-sdk-macos-x86_64.tar.bz2',
+  linux: 'https://builds.tessel.io/t2/sdk/t2-sdk-linux-x86_64.tar.bz2',
+};
+
+var RUST_LIB_TGZ_URL = 'https://builds.tessel.io/t2/sdk/t2-rustlib-VERSION.tar.gz';
+
+// Get the platform identifier. This actually conforms to the list of OSes
+// Rust supports, not the value of process.platform, so we need to convert it.
+// See: https://doc.rust-lang.org/std/env/consts/constant.OS.html
+function getPlatform() {
+  switch (process.platform) {
+    case 'darwin':
+      return 'macos';
+    case 'linux':
+      return 'linux';
+    default:
+      throw new Error('Your platform is not yet supported for cross-compilation.');
+  }
+}
+
+function sha256stream() {
+  var sha256 = createHash('sha256');
+  var stream = new Transform();
+  stream._transform = function(chunk, encoding, callback) {
+    this.push(chunk);
+    sha256.update(chunk);
+    callback();
+  };
+  stream.on('finish', () => {
+    stream.emit('sha256', sha256.digest('hex'));
+  });
+  return stream;
+}
+
+function sha256file(hash, name) {
+  return `${hash}  ${name}\n`;
+}
+
+function download(url) {
+  var req = request.get(url);
+
+  // When we receive the response
+  req.on('response', (res) => {
+
+    // Parse out the length of the incoming bundle
+    var contentLength = parseInt(res.headers['content-length'], 10);
+
+    // Create a new progress bar
+    var bar = new Progress('     [:bar] :percent :etas remaining', {
+      clear: true,
+      complete: '=',
+      incomplete: ' ',
+      width: 20,
+      total: contentLength
+    });
+
+    // When we get incoming data, update the progress bar
+    res.on('data', (chunk) => {
+      bar.tick(chunk.length);
+    });
+  });
+
+  return req;
+}
+
+function downloadString(url) {
+  return new Promise((resolve, reject) => {
+    request({
+      url,
+      // We want to force Cloudfront to serve us the latest file.
+      headers: {
+        'Accept-Encoding': 'gzip, deflate',
+      },
+    }, (error, response, body) => {
+      if (!error && response.statusCode === 200) {
+        resolve(body);
+      } else {
+        reject(error || response.statusCode);
+      }
+    });
+  });
+}
+
+function tmpdir() {
+  return new Promise((resolve) => {
+    var dir = fsTemp.template('t2-sdk-%s').mkdirSync();
+    resolve({
+      path: dir,
+      cleanup: () => {
+        try {
+          fs.removeSync(dir);
+        } catch (e) {
+          // Swallow errors in removing temporary folder. If the folder was
+          // successfully or unsuccessfully moved, it may not exist at its
+          // location in the temporary directory, but this isn't fatal.
```

Comment:
(Note that the move *should* be atomic, but fs.move allows us gracefully to move across disjoint filesystems, and I assume that is a non-atomic, possibly copy operation.)